### DISCUSSION TITLE
Tracking open/close connections efficiently?

### DISCUSSION THREAD

**Discussion by cnicodeme** [aio-libs/aiosmtpd#290] [Q&A]
Hi!

(This is not an issue, sorry if it's not i the right place)

I'm wondering if there is a way to track when a user is connecting, but also when disconnecting, with guarantee about the disconnect (connection closed, lost, exception, error, etc).

The idea is to implement a counter to track the number of active connections at a specific time to limit the abusers.

Thanks in advance :)

**Comment by waynew** [comment]
We didn't have discussions enabled before, but I figure they would be a good choice for our project... and this issue is perfect for a discussion :joy: 

---

I don't know for sure if the functionality is built-in in a friendly way, but I know that I'm tracking it via [session.host_name](https://github.com/waynew/orouboros/blob/4746fb2fb805283321071fe1bb1be11d8380c755/orouboros.py#L202)

I'm pretty sure that you can get the client connection from somewhere. You could check the aiosmtpd code to see how the hostname gets passed - if that's coming directly from the client socket then it's at least as trustworthy as any other mechanism. But if you have a proxy in front of your server then it *might* report the proxy as the client, rather than the ultimate end client, so you may need to do some poking around to figure it out. But generally speaking I *think* that the session object should, ya know, contain each session.

HTH!

**Comment by cnicodeme** [comment]
Wonderful :D !

I can easily track the "connections" part with the `handle_HELO` / `handle_EHLO` part, but it's more related to the disconnect where I'm having troubles.

Not every server will do a proper `QUIT`, and even if they would, any arising exceptions on my end could close the connection prematurely and cause the counter to not decrease.

One thing I thought about was overriding the `connection_lost()` from smtpd.py, but I'm not entirely sure it would be called every time with certainty. Do you know if that is the case ?

You mention tracking the session object. I agree that is an interesting idea, but the same remain : how to track that the session object is "closed" (set to none) ?

Another way, maybe (again, not sure), would be to somehow track the `self._handler_coroutine` event and add a callback on cancel?

**Comment by cnicodeme** [comment]
Trying to find the best solution, I came up with something that seems robust: I add a "done callback" to the handler_coroutine :

`server._handler_coroutine.add_done_callback(some_func)`

From my tests, it seems to handle all the cases (normal quit, disconnect from server, from client, exception, hard close).

Of course, I'm open to a better and more robust solution if any !

**Comment by pepoluan** [comment]
Seems to be a good feature to be added publicly (no leading underscore) in 1.5.0 :)
